File: 0102-Binary-Tree-Level-Order-Traversal/src/Solution.java
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

/**
 * @author youyu.song
 * @date 2019/9/21 9:15
 */
public class Solution {

	public class TreeNode {
		int val;
		TreeNode left;
		TreeNode right;

		TreeNode(int x) {
			val = x;
		}
	}

	/**
	 * 给定一个二叉树，返回其按层次遍历的节点值。 （即逐层地，从左到右访问所有节点）。
	 *
	 * 例如:
	 * 给定二叉树: [3,9,20,null,null,15,7],
	 *     3
	 *    / \
	 *   9  20
	 *     /  \
	 *    15   7
	 * 返回其层次遍历结果：
	 * [
	 *   [3],
	 *   [9,20],
	 *   [15,7]
	 * ]
	 *
	 * 来源：力扣（LeetCode）
	 * 链接：https://leetcode-cn.com/problems/binary-tree-level-order-traversal
	 *
	 * @param root
	 * @return
	 */
	public List<List<Integer>> levelOrder(TreeNode root) {
		if (root == null) {
			return new ArrayList<>();
		}

		List<List<Integer>> result = new ArrayList<>();
		Queue<TreeNode> queue = new LinkedList<>();
		queue.add(root);

		while (! queue.isEmpty()) {
			List<Integer> list = new ArrayList<>();
			int size = queue.size();
			for (int i = 0; i < size; i ++) {
				TreeNode treeNode = queue.poll();
				list.add(treeNode.val);
				if (treeNode.left != null) {
					queue.add(treeNode.left);
				}
				if (treeNode.right != null) {
					queue.add(treeNode.right);
				}
			}
			result.add(list);
		}

		return result;
	}

}

File: 0102-Binary-Tree-Level-Order-Traversal/Python/Solution.py
from typing import Optional, List, Deque
from collections import deque


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if root is None:
            return []
        
        result: List[List[int]] = []
        q: Deque[TreeNode] = deque()
        q.append(root)

        while q:
            size: int = len(q)
            sub_list: List[int] = []
            for _i in range(size):
                tn = q.popleft()
                sub_list.append(tn.val)
                if tn.left:
                    q.append(tn.left)
                if tn.right:
                    q.append(tn.right)

            result.append(sub_list)

        return result
    

    def levelOrder_1(self, root: Optional[TreeNode]) -> List[List[int]]:
        if root is None:
            return []
        
        result: List[List[int]] = []
        self.dfs(root, 0, result)

        return result
    

    def dfs(self, root: Optional[TreeNode], level: int, result: List[List[int]]) -> None:
        if root is None:
            return
        
        if level > len(result) - 1:
            sub_list: List[int] = []
            result.append(sub_list)

        result[level].append(root.val)    
        self.dfs(root.left, level + 1, result)
        self.dfs(root.right, level + 1, result)
            
            

            

            
